File: Baseline(AverageFill)/run_tp.m
clc; close all; clear;

addpath('src');
logger('===============================');
logger('Baseline: Average filling (AF).');
logger('===============================');

if exist('resultFolder', 'dir')==0
    system('mkdir resultFolder');
end

% load the dataset
dataPath = '../data/rawData/tpMatrix.txt';
logger(sprintf('Load data: %s', dataPath));
dataMatrix = load(dataPath);

% * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
% config area density = [0.01 : 0.01 : 0.05, 0.1 : 0.05 : 0.5];
rounds = 20;
density = [0.45 : 0.05 : 0.5];
saveTimeInfo = false; % whether to keep track of the running time
% * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *

% run for each density
runTime = tic; % start timing
for i = 1 : length(density)
    dataFolder = sprintf('../data/generatedData/TP/density_%.2f', density(i));
    outPath = sprintf('resultFolder/tpResult_%.2f.txt', density(i));
    averagefill(dataMatrix, dataFolder, outPath, density(i), rounds, saveTimeInfo) 
end  

logger('===============================');
logger(sprintf('All done. Total running time: %f s', toc(runTime)));
logger('===============================');
rmpath('src');

File: AverageFill/run_tp.py
########################################################
# run_rt.py 
# Author: Jamie Zhu <jimzhu@GitHub>
# Created: 2014/2/6
# Last updated: 2014/7/16
# Implemented approach: AverageFill
# Evaluation metrics: MAE, NMAE, RMSE, MRE, NPRE
########################################################

import numpy as np
import os, sys, time
import multiprocessing
sys.path.append('src')
from utilities import *
import evaluator
import dataloader
 

#########################################################
# config area
#
para = {'dataType': 'tp', # set 'tp' for throughput prediction
		'dataPath': '../data/dataset#1/',
		'outPath': 'result/',
		'metrics': ['MAE', 'NMAE', 'RMSE', 'MRE', 'NPRE'], # delete where appropriate		
		'density': list(np.arange(0.05, 0.31, 0.05)), # matrix density 
		'rounds': 20, # how many runs are performed at each matrix density
		'saveTimeInfo': False, # whether to keep track of the running time
		'saveLog': False, # whether to save log into file
		'debugMode': False, # whether to record the debug info
        'parallelMode': True # whether to leverage multiprocessing for speedup
		}

initConfig(para)
#########################################################


startTime = time.clock() # start timing
logger.info('==============================================')
logger.info('Baseline approach: AverageFill.')

# load the dataset
dataMatrix = dataloader.load(para)
logger.info('Loading data done.')

# run for each density
if para['parallelMode']: # run on multiple processes
    pool = multiprocessing.Pool()
    for density in para['density']:
		pool.apply_async(evaluator.execute, (dataMatrix, density, para))
    pool.close()
    pool.join()
else: # run on single processes
	for density in para['density']:
		evaluator.execute(dataMatrix, density, para)

logger.info(time.strftime('All done. Total running time: %d-th day - %Hhour - %Mmin - %Ssec.',
         time.gmtime(time.clock() - startTime)))
logger.info('==============================================')
sys.path.remove('src')